### DISCUSSION TITLE
NameError: name 'timers' isn't defined

### DISCUSSION THREAD

**Discussion by mdibit** [micropython/micropython#14106] [RP2040 / Pico]
To whom is reading this post : this program is just a simplified exercise to better understand the strange behavior of my original program where the thread Semaforo should be substituted by a robot code which should receive command from web app.
So :
The error generated by the code:
Unhandled exception in thread started by <function thread_Semaforo at 0x20015d80>
Traceback (most recent call last):
  File "<stdin>", line 91, in thread_Semaforo
  File "<stdin>", line 44, in timerWait
  File "<stdin>", line 34, in timerAttivo
NameError: name 'timers' isn't defined

I'd like to understand why, in the following code, I get the error only if the function generating the error is called from the thread Semaforo but it's perfectly working if called from the main program (or main thread).
What is **even more strange** is that, in the same function, the error is generated after the same name has been used before. 
Thanks
Michele
------------------------------------------------
import network
import socket
import time
import random
from machine import Pin
import _thread

TEMPO_STOP = 2**31
global timers
timers = [TEMPO_STOP,TEMPO_STOP,TEMPO_STOP,TEMPO_STOP,TEMPO_STOP,TEMPO_STOP]
global T_1
T_1 = 1      # timer 1

# Create an LED object on pin 'LED'
led = Pin('LED', Pin.OUT)

# Wi-Fi credentials
ssid = 'YYYYYYY'
password = 'XXXXXXXX'

pulsantePremuto = False

def timerStart(timer, durata):
    global timers
    timers[timer] = time.ticks_add(time.ticks_ms(),durata)

def timerStop(timer):
    global timers, TEMPO_STOP
    timers[timer] = TEMPO_STOP
    
def timerAttivo(timer):
    global timers, TEMPO_STOP
    return (timers[timer] != TEMPO_STOP)
    
def timerScaduto(timer):
    global timers
    return ((time.ticks_ms() > timers[timer]) and timerAttivo(timer))

def timerWait(durata):
    global T_1
    timerStart(T_1, durata)  
    while ( timerAttivo(T_1)):
        pass
    timerStop(T_1)

# HTML template for the webpage
def webpage(random_value, state):
    html = f"""
        <!DOCTYPE html>
        <html>
        <head>
            <title>Pico Web Server</title>
            <meta name="viewport" content="width=device-width, initial-scale=1">
        </head>
        <body>
            <h1>Raspberry Pi Pico Web Server</h1>
            <h2>Led Control</h2>
            <form action="./lighton">
                <input type="submit" value="Light on" />
            </form>
            <br>
            <form action="./lightoff">
                <input type="submit" value="Light off" />
            </form>
            <p>LED state: {state}</p>
            <h2>Fetch New Value</h2>
            <form action="./value">
                <input type="submit" value="Fetch value" />
            </form>
            <p>Fetched value: {random_value}</p>
        </body>
        </html>
        """
    return str(html)

def thread_Semaforo():
    global pulsantePremuto
    global T_1
    while (True):
        print("Rosso")
        if pulsantePremuto == True:
            for i in range(10):
                led.value(1)
                print("ON",i)
                timerWait(1000)
                led.value(0)
                print("OFF",i)
                timerWait(1000)
            #pulsantePremuto = False
        timerWait(5000)
        print("Verde")
        timerWait(5000)
        print("Giallo")
        timerWait(2000)

#-------------------------------------------------------------------------------------

_thread.start_new_thread(thread_Semaforo, ())

print("Connect to WLAN")
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect(ssid, password)

# Wait for Wi-Fi connection
connection_timeout = 10
while connection_timeout > 0:
    if wlan.status() >= 3:
        break
    connection_timeout -= 1
    print('Waiting for Wi-Fi connection...')
    time.sleep(1)

# Check if connection is successful
if wlan.status() != 3:
    raise RuntimeError('Failed to establish a network connection')
else:
    print('Connection successful!')
    network_info = wlan.ifconfig()
    print('IP address:', network_info[0])

# Set up socket and start listening
addr = socket.getaddrinfo('0.0.0.0', 80)[0][-1]
s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(addr)
s.listen()

print('Listening on', addr)

# Initialize variables
state = "OFF"
random_value = 0

# Main loop to listen for connections
while True:
    try:
        print("qui")
        conn, addr = s.accept()
        print('Got a connection from', addr)
        
        # Receive and parse the request
        request = conn.recv(1024)
        request = str(request)
        print('Request content = %s' % request)

        try:
            request = request.split()[1]
            print('Request:', request)
        except IndexError:
            pass
        
        # Process the request and update variables
        if request == '/lighton?':
            print("LED on")
            led.value(1)
            state = "ON"
        elif request == '/lightoff?':
            led.value(0)
            state = 'OFF'
            pulsantePremuto = False
        elif request == '/value?':
            random_value = random.randint(0, 20)
            pulsantePremuto = True

        # Generate HTML response
        response = webpage(random_value, state)  

        # Send the HTTP response and close the connection
        conn.send('HTTP/1.0 200 OK\r\nContent-type: text/html\r\n\r\n')
        conn.send(response)
        conn.close()
        print("qua")

    except OSError as e:
        conn.close()
        print('Connection closed')

**Comment by mdibit** [comment]
I'm sorry for the code not well indented, I'll try to get it better (maybe with mark down)

**Reply by robert-hh** [reply]
See the instructions. For code use lines with triple backticks before and below the code.
